back-nest/src/ws.gateway.ts

```diff
@@ -0,0 +1,84 @@
+import {
+  WebSocketGateway,
+  WebSocketServer,
+  SubscribeMessage,
+  MessageBody,
+} from '@nestjs/websockets';
+import { Logger } from '@nestjs/common';
+import { Server } from 'socket.io';
+
+import {
+  ScraperServiceFactory,
+  ServiceType,
+} from '../src/sсrapers/ScraperServiceFactory';
+import { PrismaService } from '../prisma/prisma.service';
+
+enum SourceType {
+  telemart = 'TELEMART',
+  rozetka = 'ROZETKA',
+}
+
+@WebSocketGateway({ cors: true })
+export class ScraperGateway {
+  @WebSocketServer() server: Server;
+  private readonly logger = new Logger(ScraperGateway.name);
+  private isScraping = false;
+
+  constructor(
+    private readonly scraperServiceFactory: ScraperServiceFactory,
+    private readonly prisma: PrismaService,
+  ) {}
+
+  @SubscribeMessage('startScraping')
+  async handleScraping(@MessageBody() serviceType: string) {
+    if (this.isScraping) {
+      this.logger.warn(`Scraper is already running.`);
+      // this.server.emit('scrapingStatus', {
+      //   message: `Another scraper is already running`,
+      // });
+      return;
+    }
```

Comment:
You can use [redlock](https://www.npmjs.com/package/redlock) for locking requests